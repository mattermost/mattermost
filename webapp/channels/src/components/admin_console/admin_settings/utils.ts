// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AdminConfig, EnvironmentConfig} from '@mattermost/types/config';

function getConfigValue(config: AdminConfig | Partial<EnvironmentConfig>, path: string) {
    const pathParts = path.split('.');

    return pathParts.reduce((obj: object | null, pathPart) => {
        if (!obj) {
            return null;
        }
        return obj[(pathPart as keyof object)];
    }, config);
}

export function isSetByEnv(environmentConfig: Partial<EnvironmentConfig>, path: string) {
    return Boolean(getConfigValue(environmentConfig, path));
}

export const parseIntNonNegative = (str: string | number, defaultValue?: number) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        if (defaultValue) {
            return defaultValue;
        }
        return 0;
    }

    return n;
};

export const parseIntZeroOrMin = (str: string | number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < 0) {
        return 0;
    }
    if (n > 0 && n < minimumValue) {
        return minimumValue;
    }

    return n;
};

export const parseIntNonZero = (str: string | number, defaultValue?: number, minimumValue = 1) => {
    const n = typeof str === 'string' ? parseInt(str, 10) : str;

    if (isNaN(n) || n < minimumValue) {
        if (defaultValue) {
            return defaultValue;
        }
        return 1;
    }

    return n;
};
