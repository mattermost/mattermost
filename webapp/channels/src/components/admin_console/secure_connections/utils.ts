// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {LocationDescriptor} from 'history';
import {DateTime, Interval} from 'luxon';
import {useCallback, useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {ClientError} from '@mattermost/client';
import type {Channel} from '@mattermost/types/channels';
import {isRemoteClusterPatch, type RemoteCluster, type RemoteClusterPatch} from '@mattermost/types/remote_clusters';
import type {SharedChannelRemote} from '@mattermost/types/shared_channels';
import type {Team} from '@mattermost/types/teams';
import type {IDMappedObjects, RelationOneToOne} from '@mattermost/types/utilities';

import {ChannelTypes} from 'mattermost-redux/action_types';
import {getChannel as fetchChannel} from 'mattermost-redux/actions/channels';
import {Client4} from 'mattermost-redux/client';
import {getChannel} from 'mattermost-redux/selectors/entities/channels';
import {getActiveTeamsList, getTeam} from 'mattermost-redux/selectors/entities/teams';

import type {ActionFuncAsync} from 'types/store';

export const useRemoteClusters = () => {
    const [remoteClusters, setRemoteClusters] = useState<RemoteCluster[]>();
    const [loadingState, setLoadingState] = useState<boolean | ClientError>(true);
    const {loading, error} = loadingStatus(loadingState);

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getRemoteClusters({excludePlugins: true});
            setRemoteClusters(data?.length ? data : undefined);
            setLoadingState(false);
        } catch (err) {
            setLoadingState(err);
        }
    };

    useEffect(() => {
        fetch();
    }, []);

    return [remoteClusters, {loading, fetch, error}] as const;
};

export const useRemoteClusterEdit = (remoteId: string | 'create', initRemoteCluster: RemoteCluster | undefined) => {
    const editing = remoteId !== 'create';

    const [currentRemoteCluster, setCurrentRemoteCluster] = useState<RemoteCluster | undefined>(initRemoteCluster);
    const [patch, setPatch] = useState<Partial<RemoteClusterPatch>>({});
    const [loading, setLoading] = useState<TLoadingState>(editing && !currentRemoteCluster);
    const [saving, setSaving] = useState<TLoadingState>(false);

    const hasChanges = Object.keys(patch).length > 0;

    useEffect(() => {
        if (!editing) {
            return;
        }
        (async () => {
            try {
                const data = await Client4.getRemoteCluster(remoteId);
                setCurrentRemoteCluster(data);
                setLoading(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
        })();
    }, [remoteId]);

    const applyPatch = (patch: Partial<RemoteClusterPatch>) => {
        setPatch((current) => ({...current, ...patch}));
    };

    const save = async () => {
        if (currentRemoteCluster && isRemoteClusterPatch(patch)) {
            setSaving(true);
            try {
                const data = await Client4.patchRemoteCluster(remoteId, patch);
                setCurrentRemoteCluster(data);
                setSaving(false);
                setPatch({});
            } catch (err) {
                setSaving(err);
            }
            setSaving(false);
        }
    };

    return [
        {...currentRemoteCluster, ...patch},
        {
            applyPatch,
            save,
            hasChanges,
            loading,
            saving,
            currentRemoteCluster,
            patch,
        },
    ] as const;
};

export const useSharedChannelRemotes = (remoteId: string) => {
    const [remotes, setRemotes] = useState<RelationOneToOne<Channel, SharedChannelRemote>>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = async () => {
        setLoadingState(true);
        try {
            const data = await Client4.getSharedChannelRemotes(remoteId, {include_deleted: true, include_unconfirmed: true});

            setRemotes(data?.reduce<typeof remotes>((state, remote) => {
                state![remote.channel_id] = remote;

                return state;
            }, {}));
            setLoadingState(false);
        } catch (error) {
            setLoadingState(error);
        }
    };

    useEffect(() => {
        fetch();
    }, [remoteId]);

    return [remotes, {loading, error, fetch}] as const;
};

const remoteRow = (remote: SharedChannelRemote, channel: Channel, team?: Team) => {
    return {...remote, display_name: channel.display_name, team_display_name: team?.display_name ?? ''};
};

export type SharedChannelRemoteRow = SharedChannelRemote & Pick<Channel, 'display_name'> & {team_display_name: string};
export const useSharedChannelRemoteRows = (remoteId: string, opts: {filter: 'home' | 'remote' | undefined}) => {
    const [sharedChannelRemotes, setSharedChannelRemotes] = useState<SharedChannelRemoteRow[]>();
    const [loadingState, setLoadingState] = useState<TLoadingState>(true);
    const dispatch = useDispatch();

    const loading = isPendingState(loadingState);
    const error = !loading && loadingState;

    const fetch = useCallback(async () => {
        if (opts.filter === undefined) {
            // wait for a filter
            return;
        }

        setLoadingState(true);
        dispatch<ActionFuncAsync<IDMappedObjects<SharedChannelRemoteRow>>>(async (dispatch, getState) => {
            const collected: IDMappedObjects<SharedChannelRemoteRow> = {};
            const missing: SharedChannelRemote[] = [];

            try {
                const data = await Client4.getSharedChannelRemotes(remoteId, {include_unconfirmed: true, exclude_remote: opts.filter === 'home', exclude_home: opts.filter === 'remote'});
                let state = getState();
                let getMyChannelsOnce: undefined | ((remote: SharedChannelRemote) => Promise<Channel | undefined>) = async (firstNotFound: SharedChannelRemote) => {
                    const channels = await Client4.getAllTeamsChannels();
                    dispatch({
                        type: ChannelTypes.RECEIVED_ALL_CHANNELS,
                        data: channels,
                    });
                    state = getState();

                    getMyChannelsOnce = undefined; // once-only

                    // return triggering not-found remote
                    return getChannel(state, firstNotFound.channel_id);
                };

                // first-past, get known channels or do initial load on first not-found channel
                for (const remote of data) {
                    // eslint-disable-next-line no-await-in-loop
                    const channel = getChannel(state, remote.channel_id) ?? await getMyChannelsOnce?.(remote);

                    if (!channel) {
                        // collect all remotes with missing channels
                        missing?.push(remote);
                        continue;
                    }

                    const team = getTeam(state, channel.team_id);
                    collected[remote.id] = remoteRow(remote, channel, team);
                }

                // fetch missing channels individually
                if (missing.length) {
                    // TODO: performance; consider adding sharedchannelremotes search param to api/v4/channels
                    await Promise.allSettled(missing.map((remote) => dispatch(fetchChannel(remote.channel_id))));
                    state = getState();

                    for (const remote of missing) {
                        const channel = getChannel(state, remote.channel_id);

                        if (!channel) {
                            continue;
                        }

                        const team = getTeam(state, channel.team_id);
                        collected[remote.id] = remoteRow(remote, channel, team);
                    }
                }

                const rows = Object.values(collected);

                setSharedChannelRemotes(rows.length ? rows : undefined);
                setLoadingState(false);
            } catch (error) {
                setLoadingState(error);
                return {error};
            }

            return {data: collected};
        });
    }, [remoteId, opts.filter]);

    useEffect(() => {
        fetch();
    }, [remoteId, opts.filter]);

    return [sharedChannelRemotes, {loading, error, fetch}] as const;
};

export const useTeamOptions = () => {
    const teams = useSelector(getActiveTeamsList);
    const teamsById = useMemo(() => teams.reduce<IDMappedObjects<Team>>((teams, team) => ({...teams, [team.id]: team}), {}), [teams]);
    return teamsById;
};

export const getEditLocation = (rc: RemoteCluster): LocationDescriptor<RemoteCluster> => {
    return {pathname: `/admin_console/site_config/secure_connections/${rc.remote_id}`, state: rc};
};

export const getCreateLocation = (): LocationDescriptor<RemoteCluster> => {
    return {pathname: '/admin_console/site_config/secure_connections/create'};
};

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));

export type TLoadingState<TError extends Error = ClientError> = boolean | TError;
export const isPendingState = <T extends Error>(loadingState: TLoadingState<T>) => loadingState === true;
export const isErrorState = <T extends Error>(loadingState: TLoadingState<T>): loadingState is T => loadingState instanceof Error;

const loadingStatus = <T extends Error>(loadingState: TLoadingState<T>) => {
    const loading = isPendingState(loadingState);
    const error = isErrorState(loadingState) ? loadingState : undefined;

    return {error, loading};
};
