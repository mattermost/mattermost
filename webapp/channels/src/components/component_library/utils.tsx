// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

function buildPropsLists(inputPossibilities: {[x: string]: any[]}): Array<{[x: string]: any}> {
    const keys = Object.keys(inputPossibilities);
    if (!keys.length) {
        return [{}];
    }

    const selectedKey = keys[0];
    const restPossibilities = {...inputPossibilities};
    delete restPossibilities[selectedKey];
    const subProps = buildPropsLists(restPossibilities);
    const result: Array<{[x: string]: any}> = [];
    inputPossibilities[selectedKey].forEach((v) => {
        subProps.forEach((rest) => {
            result.push({...rest, [selectedKey]: v});
        });
    });

    return result;
}

function buildPropString(inputProps: {[x: string]: any}) {
    const propKeys = Object.keys(inputProps);
    if (!propKeys.length) {
        return undefined;
    }

    const result = [(<>{'PROPS: '}</>)];
    propKeys.forEach((v) => {
        result.push((<><b>{v}</b>{`: ${inputProps[v]}, `}</>));
    });
    return result;
}

export function buildComponent(
    Component: React.ComponentType<any>,
    propPossibilities: {[x: string]: any[]},
    dropdownPossibilities: Array<{[x: string]: string[]} | undefined>,
    setProps: Array<{[x: string]: any} | undefined>,
) {
    const res: React.ReactNode[] = [];
    let currentPropPossibilities = {...propPossibilities};
    dropdownPossibilities.forEach((v) => {
        if (v) {
            currentPropPossibilities = {
                ...currentPropPossibilities,
                ...v,
            };
        }
    });

    const propsVariations = buildPropsLists(currentPropPossibilities);
    let builtSetProps = {};
    setProps.forEach((v) => {
        if (v) {
            builtSetProps = {
                ...builtSetProps,
                ...v,
            };
        }
    });
    propsVariations.forEach((v) => {
        const propString = buildPropString(v);
        res.push(
            <>
                {Boolean(propString) && <p>{propString}</p>}
                <Component
                    {...builtSetProps}
                    {...v}
                />
            </>,
        );
    });
    return res;
}
