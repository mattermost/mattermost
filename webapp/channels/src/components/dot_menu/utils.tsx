// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type React from 'react';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (): void => {
    // Telemetry tracking removed
};
