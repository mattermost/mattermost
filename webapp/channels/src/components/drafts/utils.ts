// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {IntlShape} from 'react-intl';
import {defineMessages} from 'react-intl';

import type {ScheduledPostErrorCode} from '@mattermost/types/schedule_post';

const errorCodeToErrorMessage = defineMessages<ScheduledPostErrorCode>({
    unknown: {
        id: 'scheduled_post.error_code.unknown_error',
        defaultMessage: 'Unknown Error',
    },
    channel_archived: {
        id: 'scheduled_post.error_code.channel_archived',
        defaultMessage: 'Channel Archived',
    },
    channel_not_found: {
        id: 'scheduled_post.error_code.channel_removed',
        defaultMessage: 'Channel Removed',
    },
    user_missing: {
        id: 'scheduled_post.error_code.user_missing',
        defaultMessage: 'User Deleted',
    },
    user_deleted: {
        id: 'scheduled_post.error_code.user_deleted',
        defaultMessage: 'User Deleted',
    },
    no_channel_permission: {
        id: 'scheduled_post.error_code.no_channel_permission',
        defaultMessage: 'Missing Permission',
    },
    no_channel_member: {
        id: 'scheduled_post.error_code.no_channel_member',
        defaultMessage: 'Not In Channel',
    },
    thread_deleted: {
        id: 'scheduled_post.error_code.thread_deleted',
        defaultMessage: 'Thread Deleted',
    },
    unable_to_send: {
        id: 'scheduled_post.error_code.unable_to_send',
        defaultMessage: 'Unable to Send',
    },
    invalid_post: {
        id: 'scheduled_post.error_code.invalid_post',
        defaultMessage: 'Invalid Post',
    },
});

export function getErrorStringFromCode(intl: IntlShape, errorCode: ScheduledPostErrorCode = 'unknown') {
    const textDefinition = errorCodeToErrorMessage[errorCode] ?? errorCodeToErrorMessage.unknown;
    return intl.formatMessage(textDefinition).toUpperCase();
}
