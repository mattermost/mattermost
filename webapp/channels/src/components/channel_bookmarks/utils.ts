// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useEffect, useMemo, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import type {Channel} from '@mattermost/types/channels';
import type {GlobalState} from '@mattermost/types/store';

import {Permissions} from 'mattermost-redux/constants';
import {getChannelBookmarks} from 'mattermost-redux/selectors/entities/channel_bookmarks';
import {getChannel, getMyChannelMember} from 'mattermost-redux/selectors/entities/channels';
import {getConfig, getFeatureFlagValue, getLicense} from 'mattermost-redux/selectors/entities/general';
import {haveIChannelPermission} from 'mattermost-redux/selectors/entities/roles';
import {insertWithoutDuplicates} from 'mattermost-redux/utils/array_utils';

import {fetchChannelBookmarks, reorderBookmark} from 'actions/channel_bookmarks';
import {loadCustomEmojisIfNeeded} from 'actions/emoji_actions';

import Constants from 'utils/constants';
import {trimmedEmojiName} from 'utils/emoji_utils';
import {canUploadFiles, isPublicLinksEnabled} from 'utils/file_utils';

export const MAX_BOOKMARKS_PER_CHANNEL = 50;

const {OPEN_CHANNEL, PRIVATE_CHANNEL, GM_CHANNEL, DM_CHANNEL} = Constants as {OPEN_CHANNEL: 'O'; PRIVATE_CHANNEL: 'P'; GM_CHANNEL: 'G'; DM_CHANNEL: 'D'};

type TAction = 'add' | 'edit' | 'delete' | 'order';
type TActionKey = `${TAction}${typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL}`;

const key = (a: TAction, c: typeof OPEN_CHANNEL | typeof PRIVATE_CHANNEL): TActionKey => {
    return `${a}${c}`;
};

const BOOKMARK_PERMISSION = {

    // open channel
    [key('add', OPEN_CHANNEL)]: Permissions.ADD_BOOKMARK_PUBLIC_CHANNEL,
    [key('edit', OPEN_CHANNEL)]: Permissions.EDIT_BOOKMARK_PUBLIC_CHANNEL,
    [key('delete', OPEN_CHANNEL)]: Permissions.DELETE_BOOKMARK_PUBLIC_CHANNEL,
    [key('order', OPEN_CHANNEL)]: Permissions.ORDER_BOOKMARK_PUBLIC_CHANNEL,

    // private channel
    [key('add', PRIVATE_CHANNEL)]: Permissions.ADD_BOOKMARK_PRIVATE_CHANNEL,
    [key('edit', PRIVATE_CHANNEL)]: Permissions.EDIT_BOOKMARK_PRIVATE_CHANNEL,
    [key('delete', PRIVATE_CHANNEL)]: Permissions.DELETE_BOOKMARK_PRIVATE_CHANNEL,
    [key('order', PRIVATE_CHANNEL)]: Permissions.ORDER_BOOKMARK_PRIVATE_CHANNEL,
} as const;

export const useChannelBookmarkPermission = (channelId: string, action: TAction) => {
    return useSelector((state: GlobalState) => getHaveIChannelBookmarkPermission(state, channelId, action));
};

export const getHaveIChannelBookmarkPermission = (state: GlobalState, channelId: string, action: TAction) => {
    const channel: Channel | undefined = getChannel(state, channelId);

    if (!channel) {
        return false;
    }

    if (channel.delete_at !== 0) {
        return false;
    }

    const {type} = channel;

    if (type === 'threads') {
        return false;
    }

    if (type === GM_CHANNEL || type === DM_CHANNEL) {
        const myMembership = getMyChannelMember(state, channelId);
        return myMembership?.channel_id === channelId;
    }

    const permission = BOOKMARK_PERMISSION[key(action, type)];

    return channel && permission && haveIChannelPermission(state, channel.team_id, channelId, permission);
};

export const useCanUploadFiles = () => {
    return useSelector((state: GlobalState) => canUploadFiles(getConfig(state)));
};

export const useCanGetPublicLink = () => {
    return useSelector((state: GlobalState) => isPublicLinksEnabled(getConfig(state)));
};

export const useCanGetLinkPreviews = () => {
    return useSelector((state: GlobalState) => getConfig(state).EnableLinkPreviews === 'true');
};

export const getIsChannelBookmarksEnabled = (state: GlobalState) => {
    const isEnabled = getFeatureFlagValue(state, 'ChannelBookmarks') === 'true';

    if (!isEnabled) {
        return false;
    }

    const license = getLicense(state);

    return license?.IsLicensed === 'true';
};

export const useChannelBookmarks = (channelId: string) => {
    const dispatch = useDispatch();
    const bookmarks = useSelector((state: GlobalState) => getChannelBookmarks(state, channelId));

    const order = useMemo(() => {
        return Object.keys(bookmarks).sort((a, b) => bookmarks[a].sort_order - bookmarks[b].sort_order);
    }, [bookmarks]);
    const [tempOrder, setTempOrder] = useState<typeof order>();

    useEffect(() => {
        if (tempOrder) {
            setTempOrder(undefined);
        }
    }, [order]);

    useEffect(() => {
        if (channelId) {
            dispatch(fetchChannelBookmarks(channelId));
        }
    }, [channelId]);

    useEffect(() => {
        const emojis = Object.values(bookmarks).reduce<string[]>((result, {emoji}) => {
            if (emoji) {
                result.push(trimmedEmojiName(emoji));
            }

            return result;
        }, []);

        if (emojis.length) {
            dispatch(loadCustomEmojisIfNeeded(emojis));
        }
    }, [bookmarks]);

    const reorder = async (id: string, prevOrder: number, nextOrder: number) => {
        setTempOrder(insertWithoutDuplicates(order, id, nextOrder));
        const {error} = await dispatch(reorderBookmark(channelId, id, nextOrder));

        if (error) {
            setTempOrder(undefined);
        }
    };

    return {
        bookmarks,
        order: tempOrder ?? order,
        reorder,
    } as const;
};

