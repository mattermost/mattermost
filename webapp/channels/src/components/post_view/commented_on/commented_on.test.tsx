// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import CommentedOn from 'components/post_view/commented_on/commented_on';

import {renderWithContext, screen} from 'tests/react_testing_utils';
import {TestHelper} from 'utils/test_helper';

describe('components/post_view/CommentedOn', () => {
    const user1 = TestHelper.getUserMock({
        id: 'user1',
    });
    const post1 = TestHelper.getPostMock({
        id: 'post1',
        user_id: user1.id,
        message: 'text message',
    });

    test("should render the root post's message and author", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    posts: {
                        posts: {
                            post1,
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        expect(screen.getByText(textInChildren("Commented on some-user's message: text message"))).toBeInTheDocument();
    });

    test("should render a placeholder name when the post's author isn't loaded", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    posts: {
                        posts: {
                            post1,
                        },
                    },
                },
            },
        );

        expect(screen.getByText(textInChildren("Commented on Someone's message: text message"))).toBeInTheDocument();
    });

    test("should render a placeholder when the post isn't loaded", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
        );

        expect(screen.getByText(textInChildren("Commented on Someone's message: Loadingâ€¦"))).toBeInTheDocument();
    });

    test("should render the root post's file attachments when it has no message", () => {
        const file1 = TestHelper.getFileInfoMock({id: 'file1', create_at: 1000, name: 'image.png'});
        const file2 = TestHelper.getFileInfoMock({id: 'file2', create_at: 1001, name: 'contract.doc'});

        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    files: {
                        fileIdsByPostId: {
                            post1: [file1.id, file2.id],
                        },
                        files: {
                            file1,
                            file2,
                        },
                    },
                    posts: {
                        posts: {
                            [post1.id]: {
                                ...post1,
                                message: '',
                                file_ids: [file1.id, file2.id],
                            },
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        expect(screen.getByText(textInChildren("Commented on some-user's message: image.png plus 1 other file"))).toBeInTheDocument();
    });

    test("should render the root post's props.pretext as message", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    general: {
                        config: {
                            EnablePostUsernameOverride: 'true',
                        },
                    },
                    posts: {
                        posts: {
                            [post1.id]: {
                                ...post1,
                                message: '',
                                props: {
                                    from_webhook: 'true',
                                    override_username: 'override_username',
                                    attachments: [{
                                        pretext: 'This is a pretext',
                                    }],
                                },
                            },
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        // This incorrectly uses the post author's name due to MM-63564
        expect(screen.getByText(textInChildren("Commented on some-user's message: This is a pretext"))).toBeInTheDocument();
    });

    test("should render the root post's props.title as message", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    general: {
                        config: {
                            EnablePostUsernameOverride: 'true',
                        },
                    },
                    posts: {
                        posts: {
                            [post1.id]: {
                                ...post1,
                                message: '',
                                props: {
                                    from_webhook: 'true',
                                    override_username: 'override_username',
                                    attachments: [{
                                        title: 'This is a title',
                                    }],
                                },
                            },
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        // This incorrectly uses the post author's name due to MM-63564
        expect(screen.getByText(textInChildren("Commented on some-user's message: This is a title"))).toBeInTheDocument();
    });

    test("should render the root post's props.text as message", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    general: {
                        config: {
                            EnablePostUsernameOverride: 'true',
                        },
                    },
                    posts: {
                        posts: {
                            [post1.id]: {
                                ...post1,
                                message: '',
                                props: {
                                    from_webhook: 'true',
                                    override_username: 'override_username',
                                    attachments: [{
                                        text: 'This is a text',
                                    }],
                                },
                            },
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        // This incorrectly uses the post author's name due to MM-63564
        expect(screen.getByText(textInChildren("Commented on some-user's message: This is a text"))).toBeInTheDocument();
    });

    test("should render the root post's props.fallback as message", () => {
        renderWithContext(
            <CommentedOn rootId={post1.id}/>,
            {
                entities: {
                    general: {
                        config: {
                            EnablePostUsernameOverride: 'true',
                        },
                    },
                    posts: {
                        posts: {
                            [post1.id]: {
                                ...post1,
                                message: '',
                                props: {
                                    from_webhook: 'true',
                                    override_username: 'override_username',
                                    attachments: [{
                                        fallback: 'This is fallback message',
                                    }],
                                },
                            },
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        // This incorrectly uses the post author's name due to MM-63564
        expect(screen.getByText(textInChildren("Commented on some-user's message: This is fallback message"))).toBeInTheDocument();
    });

    test('should call onCommentClick on click of text message', () => {
        const onCommentClick = jest.fn();
        renderWithContext(
            <CommentedOn
                onCommentClick={onCommentClick}
                rootId={post1.id}
            />,
            {
                entities: {
                    posts: {
                        posts: {
                            post1,
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        screen.getByText('text message').click();

        expect(onCommentClick).toHaveBeenCalledTimes(1);
    });

    test("should render the root post's overwritten username", () => {
        const webhookPost = TestHelper.getPostMock({
            id: 'webhook_post_id',
            user_id: user1.id,
            message: 'text message',
            props: {
                from_webhook: 'true',
                override_username: 'overridden_username',
            },
        });

        const post1 = TestHelper.getPostMock({
            id: 'post1',
            user_id: user1.id,
            message: 'text message',
            root_id: webhookPost.id,
        });

        renderWithContext(
            <CommentedOn
                rootId={webhookPost.id}
                enablePostUsernameOverride={true}
            />,
            {
                entities: {
                    posts: {
                        posts: {
                            post1,
                            webhook_post_id: webhookPost,
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        expect(screen.getByText(textInChildren("Commented on overridden_username's message: text message"))).toBeInTheDocument();
    });

    test("should not render the root post's overwritten username if post is not from webhook", () => {
        const webhookPost = TestHelper.getPostMock({
            id: 'webhook_post_id',
            user_id: user1.id,
            message: 'text message',
            props: {
                override_username: 'overridden_username',
            },
        });

        const post1 = TestHelper.getPostMock({
            id: 'post1',
            user_id: user1.id,
            message: 'text message',
            root_id: webhookPost.id,
        });

        renderWithContext(
            <CommentedOn
                rootId={webhookPost.id}
                enablePostUsernameOverride={true}
            />,
            {
                entities: {
                    posts: {
                        posts: {
                            post1,
                            webhook_post_id: webhookPost,
                        },
                    },
                    users: {
                        profiles: {
                            user1,
                        },
                    },
                },
            },
        );

        expect(screen.getByText(textInChildren("Commented on some-user's message: text message"))).toBeInTheDocument();
    });
});

function textInChildren(matchedText: string) {
    return (content: string, element: Element | null) => {
        const hasText = element?.textContent === matchedText;
        const childHasText = element && Array.from(element?.children).some((child) => child?.textContent === matchedText);
        return hasText && !childHasText;
    };
}
