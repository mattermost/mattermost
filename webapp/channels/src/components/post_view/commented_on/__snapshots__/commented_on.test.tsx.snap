// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/post_view/CommentedOn should match snapshot 1`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      text message
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshot 2`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      text message
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshot 3`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      <Connect(Component)
        parentPostId="post_id"
      />
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshots for post with props.fallback as message 1`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      This is fallback message
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshots for post with props.pretext as message 1`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      This is a pretext
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshots for post with props.text as message 1`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      This is a text
    </a>
  </span>
</div>
`;

exports[`components/post_view/CommentedOn should match snapshots for post with props.title as message 1`] = `
<div
  className="post__link"
  data-testid="post-link"
>
  <span>
    <MemoizedFormattedMessage
      defaultMessage="Commented on {name}'s message: "
      id="post_body.commentedOn"
      values={
        Object {
          "name": <a
            className="theme user_name"
          >
            <Memo(Connect(UserProfile))
              userId=""
            />
          </a>,
        }
      }
    />
    <a
      className="theme"
      onClick={[MockFunction]}
    >
      This is a title
    </a>
  </span>
</div>
`;
